### PR TITLE
Add categories on json list

### PR DESCRIPTION
**Pull request by danielebarbaro**
Hi @mattstauffer , first implementation of #143 could be like this:

Screenshot 2020-10-01 at 17 30 16

The json is: 

```json
{
  "Cache": [
    [
      {
        "container_id": "3aac7ef4477f",
        "names": "TO--memcached--1.6.7",
        "status": "Up 7 hours",
        "ports": "0.0.0.0:11211->11211\\/tcp"
      }
    ]
  ],
  "Database": [
    [
      {
        "container_id": "4dc197902954",
        "names": "TO--mariadb--focal",
        "status": "Up 7 hours",
        "ports": "0.0.0.0:3342->3306\\/tcp"
      }
    ],
    [
      {
        "container_id": "0daed47e14a5",
        "names": "TO--postgresql--13.0",
        "status": "Up 3 days",
        "ports": "0.0.0.0:5647->5432\\/tcp"
      }
    ]
  ],
  "Mail": [
    [
      {
        "container_id": "202e824629b1",
        "names": "TO--mailhog--v1.0.1",
        "status": "Up 7 hours",
        "ports": "0.0.0.0:8025->8025\\/tcp, 0.0.0.0:2341->1025\\/tcp"
      }
    ]
  ]
}
```

I decided to group container by category as i've done in the principal menu #88 (PR #135) but we can discuss about it.

I think that isn't useful category column in list output.

### CODE REVIEW

**Comment by danielebarbaro** [comment]
Hi @mattstauffer now PR is updated.
All tests pass.

This PR needs #157 .

**Comment by danielebarbaro** [comment]
🥺

**Comment by mattstauffer** [comment]
@danielebarbaro Whoops, sorry!

1. I *think* I'd prefer we use the lowercase version of the categories, but I'd be up to someone telling me the uppercase is better.
2. My natural inclination would be just to add it to each item in the list rather than grouping them this way. @mpociot as the first person who's talked about possibly consuming this data, what do you think?

**Comment by danielebarbaro** [comment]
hi @mpociot , what you think about this implementation?

> I think I'd prefer we use the lowercase version of the categories, but I'd be up to someone telling me the uppercase is better.

oookkeeeei  🧐 🤓

> My natural inclination would be just to add it to each item in the list rather than grouping them this way. @mpociot as the first person who's talked about possibly consuming this data, what do you think?

I prefer grouping to "select" only one group a time 💪🏻

**Comment by mattstauffer** [comment]
@danielebarbaro OK, I'm going to say definitely throw them into the main body of each entry, and give a flat response of entries.

@josecanhelp had a great idea as well: To add the category under a `meta` key, so it's separated from the rest.

Thanks!

**Comment by danielebarbaro** [comment]
Cool, if you need any help i can support you .
I noticed your last PR merges and i saw **Base Alias** && **Full Alias** , well done @josecanhelp 🥇

**Comment by mattstauffer** [comment]
@danielebarbaro If you're up to code the change, I'd love it! If you don't have time, please let me know.

**Comment by danielebarbaro** [comment]
> @danielebarbaro If you're up to code the change, I'd love it! If you don't have time, please let me know.

i'm always on the cutting edge 🏄🏼‍♂️. i will rewrite the PR with the new changes

**Comment by danielebarbaro** [comment]
Hi @mattstauffer , PR updated but i have some doubts to discuss.

Screenshot 2020-11-10 at 15 49 26


- `takeout list --json` have different output from `takeout list` it is what we want?

My opinion is to have same output (add category column on list) and with #173 filter base alias, full alias and category fields

- Full Alias and Base Alias are lowercase, category doesn't.

My opinion is to have same output :) convert category to lowercase and use ucfirst() on menu (`takeout enable`).

- behind the scenes i use container name to get the right category, i would like to use Base Alias

My opinion is: container name isn't clean to parse, base alias is "more confident" to use

**Comment by danielebarbaro** [comment]
🙋🏻‍♂️

**Comment by mattstauffer** [comment]
> takeout list --json have different output from takeout list is it what we want?
Yep!

> Full Alias and Base Alias are lowercase, category isn't.
I agree, let's strtolower the category

> behind the scenes i use container name to get the right category however, i would like to use base_alias
Makes sense!

File: app/Services/Category.php [hunk 1]
```diff
@@ -10,4 +12,30 @@ abstract class Category
...
+    public static function all(): array
+    {
+        return collect((new Services)->all())
+            ->mapWithKeys(function ($fqcn, $shortName) {
+                return [$shortName => $fqcn::category()];
+            })
+            ->toArray();
+    }
+
+    public static function get(string $serviceName): string
```
Comment (mattstauffer) [inline app/Services/Category.php:25]: This would be `fromServiceName`, right?

File: app/Services/Category.php [hunk 2]
```diff
@@ -10,4 +12,30 @@ abstract class Category
     const SEARCH = 'Search';
     const STORAGE = 'Storage';
     const TOOLS = 'Tools';
+
+    public static function all(): array
```
Comment (mattstauffer) [inline app/Services/Category.php:16]: I think this would more be a method on the Service class named `allByCategory` or something like that?

**Review by mattstauffer — CHANGES_REQUESTED** [review]

File: app/Services/Category.php (hunk 1 above)
Comment (danielebarbaro) [reply app/Services/Category.php:25]: Yep.

File: app/Services/Category.php (hunk 2 above)
Comment (danielebarbaro) [reply app/Services/Category.php]: `all()` now returns:

```php
$result = [
    'beanstalkd' => "Cache",
    'clickhouse' => "Database",
    'dynamodb' => "Cache",
    'elasticsearch' => "Search",
    'eventstoredb' => "Database",
    'expose' => "Tools",
    'influxdb' => "Database",
    'mailhog' => "Mail",
    'mariadb' => "Database",
    'meilisearch' => "Search",
    'memcached' => "Cache",
    'minio' => "Storage",
    'mongo' => "Database",
    'mssql' => "Database",
    'mysql' => "Database",
    'neo4j' => "Database",
    'postgis' => "Database",
    'postgresql' => "Database",
    'redis' => "Cache",
    'sftp' => "Storage",
    'sqs' => "Cache"
];
```

so the right name (for me) is `mapByService()`

Comment (mattstauffer) [reply app/Services/Category.php]: Sure, but I still think this is more of a list of services than a list of categories, right?

**Comment by mattstauffer** [comment]
@danielebarbaro Thanks for all your hard work on this!

File: app/Services/Category.php (hunk 2 above)
Comment (danielebarbaro) [reply app/Services/Category.php]: whoops sorry 🙈 I misunderstood!
Apologies, you're right, it is a method for Service class.

**Comment by mattstauffer** [comment]
@danielebarbaro OK, I'm finally getting back into my open source tasks. Taking a look now!

File: app/Services/Category.php [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 namespace App\Services;
 
+use App\Services;
+
 abstract class Category
```
Comment (mattstauffer) [inline app/Services/Category.php:7]: I know this has been sitting open for ages so I'm sorry to make one more request, but could you write tests for these two new methods?

Thanks so much!

**Review by mattstauffer — APPROVED** [review]

**Comment by tonysm** [comment]
Thanks @danielebarbaro for the work on this. And sorry for the long delay!

**Pull request merged**